compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/CupertinoOverscrollEffect.kt

```diff
@@ -0,0 +1,405 @@
+                // start spring animation instead
+                lastFlingUncosumedDelta = unconsumedDelta
+
+                delta - unconsumedDelta
+            }
+        }
+    }
+
+    override fun applyToScroll(
+        delta: Offset,
+        source: NestedScrollSource,
+        performScroll: (Offset) -> Offset
+    ): Offset {
+        direction = direction.combinedWith(delta.toCupertinoOverscrollDirection())
+
+        return source.toCupertinoScrollSource()?.let {
+            applyToScroll(delta, it, performScroll)
+        } ?: performScroll(delta)
+    }
+
+    override suspend fun applyToFling(
+        velocity: Velocity,
+        performFling: suspend (Velocity) -> Velocity
+    ) {
+        val availableFlingVelocity = playInitialSpringAnimationIfNeeded(velocity)
+
+        val velocityConsumedByFling = performFling(availableFlingVelocity)
+        val postFlingVelocity = availableFlingVelocity - velocityConsumedByFling
+
+        playSpringAnimation(
+            lastFlingUncosumedDelta.toFloat(),
+            postFlingVelocity.toFloat(),
+            false
+        )
+    }
+
+    private fun Offset.toCupertinoOverscrollDirection(): CupertinoOverscrollDirection {
+        val epsilon = 1E-4f
+
+        val hasXPart = abs(x) > epsilon
+        val hasYPart = abs(y) > epsilon
+
+        return if (hasXPart xor hasYPart) {
+            if (hasXPart) {
+                CupertinoOverscrollDirection.HORIZONTAL
+            } else {
+                // hasYPart != hasXPart and hasXPart is false
+                CupertinoOverscrollDirection.VERTICAL
+            }
+        } else {
+            // hasXPart and hasYPart are equal
+            CupertinoOverscrollDirection.UNKNOWN
+        }
+    }
+
+    private fun CupertinoOverscrollDirection.combinedWith(other: CupertinoOverscrollDirection): CupertinoOverscrollDirection =
+        when (this) {
+            CupertinoOverscrollDirection.UNKNOWN -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN -> CupertinoOverscrollDirection.UNKNOWN
+                CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+                CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+            }
+
+            CupertinoOverscrollDirection.VERTICAL -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN, CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+                CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+            }
+
+            CupertinoOverscrollDirection.HORIZONTAL -> when (other) {
+                CupertinoOverscrollDirection.UNKNOWN, CupertinoOverscrollDirection.HORIZONTAL -> CupertinoOverscrollDirection.HORIZONTAL
+                CupertinoOverscrollDirection.VERTICAL -> CupertinoOverscrollDirection.VERTICAL
+            }
+        }
+    private fun Velocity.toOffset(): Offset =
```

Comment:
It doesn't look related to this class. Let's place such utils in file scope

Reply:
That's the same way `ScrollingLogic` does it. It depends on internal class state and will make all calls, which are only used internally by this class more verbose.

Reply:
You mean `Velocity` explicitly? Or all other conversions as well?

Reply:
Moved Velocity<->Offset to file scope.